### PR TITLE
Scene flow

### PR DESCRIPTION
**Pull request by Ardorax**
# Scene flow class

Create scene flow by calling the class with a file path, a tensor or a ndarray.

If you have optical flow, depth at time T and T + 1 and the camera intrinsic. You can create scene flow with the class method `from_optical_flow`. It handle the creation of the occlusion mask if some parameters have one.

### CODE REVIEW

File: aloscene/scene_flow.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        Returns
+        -------
+        flipped_scene_flow : aloscene.SceneFlow
+            horizontally flipped scene flow map
+        """
+        flow_flipped = super()._hflip(**kwargs)
+        # invert x axis of flow vector
+        labels = flow_flipped.drop_children()
+        sl_x = flow_flipped.get_slices({"C": 0})
+        sl_z = flow_flipped.get_slices({"C": 2})
```
Comment (jsalotti) [inline aloscene/scene_flow.py:105]: The depth coordinate (Z) should not change during the horizontal flip.

File: aloscene/scene_flow.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        flipped_scene_flow : aloscene.SceneFlow
+            vertically flipped scene flow map
+        """
+        flow_flipped = super()._vflip(**kwargs)
+        # invert y axis of flow vector
+        labels = flow_flipped.drop_children()
+        sl_y = flow_flipped.get_slices({"C": 1})
+        sl_z = flow_flipped.get_slices({"C": 2})
+        flow_flipped[sl_y] = -1 * flow_flipped[sl_y]
+        flow_flipped[sl_z] = -1 * flow_flipped[sl_z]
```
Comment (jsalotti) [inline aloscene/scene_flow.py:125]: Same comment as for hflip. The Z coordinate should not change during vertical flip.

File: aloscene/scene_flow.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        depth: aloscene.Depth
+            The depth at T."
+        next_depth: aloscene.Depth
+            The depth at T + 1
+        intrinsic : aloscene.CameraIntrinsic
+            The intrinsic of the image at T.
+        """
+        start_vector = depth.as_points3d(camera_intrinsic=intrinsic).cpu().numpy()
+        end_vector = next_depth.as_points3d(camera_intrinsic=intrinsic).cpu().numpy()
+
```
Comment (jsalotti) [inline aloscene/scene_flow.py:68]: You should use the optical flow to know where to read the depth in `next_depth`.

**Review by jsalotti — CHANGES_REQUESTED** [review]

File: aloscene/scene_flow.py (hunk 1 above)
Comment (Ardorax) [reply aloscene/scene_flow.py]: It's fix.

File: aloscene/scene_flow.py (hunk 2 above)
Comment (Ardorax) [reply aloscene/scene_flow.py]: It"s fix.

File: aloscene/scene_flow.py (hunk 3 above)
Comment (Ardorax) [reply aloscene/scene_flow.py:68]: Temporary fix by using for loop.

File: aloscene/scene_flow.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            The optical flow at T.
+        depth: aloscene.Depth
+            The depth at T."
+        next_depth: aloscene.Depth
+            The depth at T + 1
+        intrinsic : aloscene.CameraIntrinsic
+            The intrinsic of the image at T.
+        """
+        start_vector = np.reshape(depth.as_points3d(camera_intrinsic=intrinsic).cpu().numpy(), (depth.H, depth.W, 3))
+        next_vector = np.reshape(next_depth.as_points3d(camera_intrinsic=intrinsic).cpu().numpy(), (depth.H, depth.W, 3))
```
Comment (jsalotti) [inline aloscene/scene_flow.py:67]: Just for readability :  array have a `.reshape()` methods.
`np.reshape(a, shape)` is the same as `a.reshape(shape)`

File: aloscene/scene_flow.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        flow = np.round(optical_flow.numpy(), 0).astype(int)
+
+        y_coords, x_coords = np.mgrid[: depth.H, : depth.W]
+
+        # How to handle this
+        new_x = np.clip(x_coords + flow[:, :, 0], 0, depth.W - 1)
+        new_y = np.clip(y_coords + flow[:, :, 1], 0, depth.H - 1)
+
+        final_points = next_vector[new_y, new_x, :] - start_vector
```
Comment (jsalotti) [inline aloscene/scene_flow.py:77]: final_points is not a good names, because it's already a sceneflow vector (difference of two 3D points)

File: aloscene/scene_flow.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        scene_flow_vector = np.transpose(final_points, (2, 0, 1))
+        masked_points = Mask(np.isfinite(scene_flow_vector).all(0), names=("H", "W"))
+
+        result = torch.from_numpy(scene_flow_vector)
```
Comment (jsalotti) [inline aloscene/scene_flow.py:82]: not necessary because `AugmentedTensor` accept a numpy array as parameter

File: aloscene/scene_flow.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            raise ValueError(
+                "The optical flow, depth and next_depth must have the same names"
+            )
+
+        if optical_flow.names != ("C", "H", "W") and optical_flow.names != (
+            "B",
+            "C",
+            "H",
+            "W",
+        ):
```
Comment (jsalotti) [inline aloscene/scene_flow.py:73]: remove the last comma in `("B", "C", "H", "W",)` to avoid newlines in auto-formatting

**Review by jsalotti — CHANGES_REQUESTED** [review]

File: aloscene/scene_flow.py [hunk 8]
```diff
@@ -80,50 +80,53 @@ def from_optical_flow(
...
         start_vector = (
             depth.as_points3d(camera_intrinsic=intrinsic)
-            .cpu()
-            .numpy()
+            .as_tensor()
             .reshape((depth.H, depth.W, 3))
+            .unsqueeze(0)
+            .transpose(3,1)
+            .transpose(2,3)
         )
```
[the remark this answers is not in the document]
Comment (jsalotti) [reply aloscene/scene_flow.py:91]: note : `unsqueeze(0)` will not be necessary because you will reshape directly to the correct shape

[the remark this answers is not in the document]
Comment (Ardorax) [reply aloscene/scene_flow.py]: FIX
torch.Tensor.transpose take two dimensions as parameters.

File: aloscene/scene_flow.py [hunk 9]
```diff
@@ -80,50 +80,53 @@ def from_optical_flow(
...
-            + next_vector[floor_new_y, ceil_new_x, :]
-            + next_vector[ceil_new_y, floor_new_x, :]
-            + next_vector[floor_new_y, floor_new_x, :]
-        ) / 4 - start_vector  # type: ignore
+        flow = torch.tensor(flow)
 
-        # Transpose the dimmentions to match the pytorch convention
-        scene_flow_vector = np.transpose(scene_flow_vector, (0, 3, 1, 2))
+        # Move all the element of next_vector to the correct position
+        r = F.grid_sample(
```
[the remark this answers is not in the document]
Comment (Ardorax) [reply aloscene/scene_flow.py]: FIX

File: aloscene/scene_flow.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+        # Compute the scene flow
+        scene_flow_vector = end_vector - start_vector
+
+        # Create the occlusion mask
+        occlusion = torch.zeros(H, W, dtype=torch.bool)
+
+        # Add depth and optical flow occlusion to main occlusion mask
+        if optical_flow.occlusion is not None:
+            occlusion = occlusion | optical_flow.occlusion.as_tensor().bool() == 1
```
Comment (jsalotti) [inline aloscene/scene_flow.py:101]: what is the purpose of == 1 if you are alreay doing boolean operation on boolean tensors ?

File: aloscene/scene_flow.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        if depth.occlusion is not None:
+            occlusion = occlusion | depth.occlusion.as_tensor().bool() == 1
+        if next_depth.occlusion is not None:
+            occlusion = occlusion | next_depth.occlusion.as_tensor().bool() == 1
```
Comment (jsalotti) [inline aloscene/scene_flow.py:105]: do not add the occlusion of next_depth here; you already use it later by sampling this occlusion at the correct coordinates to create `moved_occlusion`

File: aloscene/scene_flow.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        new_y = new_y / H * 2 - 1
+        new_coords = torch.stack([new_x, new_y], dim=3)
+
+        # Move the point cloud at T + 1 to the new position
+        end_vector = F.grid_sample(next_vector, new_coords, mode=mode, padding_mode="zeros", align_corners=True)
+
+        # Compute the scene flow
+        scene_flow_vector = end_vector - start_vector
+
+        # Create the occlusion mask
```
Comment (jsalotti) [inline aloscene/scene_flow.py:96]: Check that at least one of the augmented tensors has occlusion (optical_flow, depth, next_depth).
If it is not the case, the resulting `SceneFlow` object should have `occlusion=None`

File: aloscene/scene_flow.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        occlusion = occlusion.unsqueeze(1)
+
+        # Use of 'not' needed because the grid_sample has padding_mode="zeros" and
+        # the 0 from this function mean that the pixel is occluded
+        occlusion = ~occlusion
+
+        # Move the occlusion mask like the scene flow to check if occluded pixel are used in the calculation
+        moved_occlusion = F.grid_sample(
+            occlusion.float(), new_coords, mode=mode, padding_mode="zeros", align_corners=True
+        )
```
Comment (jsalotti) [inline aloscene/scene_flow.py:116]: `moved_occlusion` should be sampled only from `next_frame.depth.occlusion`.

File: aloscene/scene_flow.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+        occlusion = ~occlusion
+        # Sometimes moved_occlusion is not exactly 1 even if the pixels around are not occluded
+        moved_occlusion = ~(moved_occlusion >= 0.99999)
+
+        # Fusion of the 2 occlusion mask
+        occlusion = occlusion | moved_occlusion == 1
```
Comment (jsalotti) [inline aloscene/scene_flow.py:122]: same as above : what is the use of `==1` ?

File: aloscene/scene_flow.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        super().__init__(x)
+
+    @classmethod
+    def from_optical_flow(
+        cls,
+        optical_flow: Flow,
+        depth: Depth,
+        next_depth: Depth,
+        intrinsic: CameraIntrinsic,
+        mode: str = "bilinear",
```
Comment (jsalotti) [inline aloscene/scene_flow.py:46]: - replace `mode` parameter by `sampling` for more explicit name. 
- add a description in the docstring

**Review by jsalotti — CHANGES_REQUESTED** [review]

File: aloscene/scene_flow.py (hunk 10 above)
Comment (Ardorax) [reply aloscene/scene_flow.py:101]: Just for a better display when debugging.

**Review by jsalotti — APPROVED** [review]

File: aloscene/scene_flow.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
+from aloception.aloscene import Depth, CameraIntrinsic, Mask, Flow
+import aloception.aloscene as aloscene
+from typing import Union
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
+def load_scene_flow(path: str) -> torch.Tensor:
```
Comment (Data-Iab) [inline aloscene/scene_flow.py:9]: You can add this function to `aloscene/io/flow.py`

File: aloscene/scene_flow.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    with open(path, "rb") as file:
+        data = np.load(file)
+        return torch.from_numpy(data)
```
Comment (Data-Iab) [inline aloscene/scene_flow.py:12]: No need to convert to tensor

File: aloscene/scene_flow.py [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+
+
+class SceneFlow(aloscene.tensors.SpatialAugmentedTensor):
+    """
+    Scene flow map
+
+    Parameters
+    ----------
+    x : str
```
Comment (Data-Iab) [inline aloscene/scene_flow.py:21]: Can also be a tensor

File: aloscene/scene_flow.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        depth: aloscene.Depth
+            The depth at T.
+        next_depth: aloscene.Depth
+            The depth at T + 1
+        intrinsic : aloscene.CameraIntrinsic
+            The intrinsic of the image at T.
+        sampling: str
+            The sampling method to use for the scene flow.
+        """
+        has_batch = True if len(optical_flow.names) == 4 else False
```
Comment (Data-Iab) [inline aloscene/scene_flow.py:63]: names can be (T, C, H, W). The correct way to check is : `has_batch = "B" in names`.
However, with the two assertions below it's correct.

**Review by Data-Iab — CHANGES_REQUESTED** [review]

File: aloscene/scene_flow.py (hunk 17 above)
Comment (Ardorax) [reply aloscene/scene_flow.py:12]: Does it mean that x can also be a numpy array if you don't convert ?

Comment (Data-Iab) [reply aloscene/scene_flow.py]: As long as your output is used to build an AugmentedTensor, you can just `return data`.

**Review by Data-Iab — APPROVED** [review]
good job

**Comment by thibo73800** [comment]
Thanks, can you add a quick/simple example on how to use it in the description ? So that I can add it to the next release ? I'll merge after
